facebookresearch/mbrl-lib
# Migration to hydra latest

## Types of changes




## Motivation and Context / Related issue





**Motivation:** Enable importing `mbrl-lib` in project requiring newer hydra-core>=1.1.2 version.
**Context:** `mbrl-lib` `hydra-core` version was freezed at [v1.0.3](https://github.com/facebookresearch/hydra/releases/tag/v1.0.3) which was release in 2020. A lot of useful functionality were introduce since then.


**Regression warning/bug that where fixed:**
Post hydra-core version 1.1 introduce breaking changes that were fixed in this PR the following way:
- `hydra-core>=1.1.0` set `instantiate` fct to recursive by default, breaking all mbrl-lib use of `hydra.instantiate` original behaviour with nested dict config ([hydra 1.1.0 release note](https://github.com/facebookresearch/hydra/releases/tag/v1.1.0)). Changing config files on the user side would be error-prone, so I simply added `_recursive_=False` to every `hydra.utils.instantiate(<cfg>)` that deal with nested instantiation logic;
- Preserve mbrl-lib default config ordering behaviour by adding `_self_` keyword in `examples/conf/main.yaml` `defaults`  (Ref https://hydra.cc/docs/1.1/upgrades/1.0_to_1.1/default_composition_order/);
- Removed deprecated package header `@package _group_` to all config group yaml file (ref https://hydra.cc/docs/1.1/upgrades/1.0_to_1.1/changes_to_package_header/).


**Remark on breaking change:** 
- overriding via command line now require prefixing with `++` eg: `python -m mbrl.examples.main algorithm=mbpo  ++device=cpu`

**Note:** 
- I've set the minimal requirement to `hydra-core>=1.1.2`. Older than that wont be compatible with the proposed changes.
- Documentation was updated


## How Has This Been Tested (if it applies)



- Repeated all the following test with hydra-core==[1.1.2, 1.2, 1.3.2]
   - Executed pytest tests on both `tests/core` and `tests/algorithms`
   - Manually run the modified example at `mbrl/examples/README.md`
      ```bash
      python -m mbrl.examples.main \
         algorithm=mbpo \
        overrides=mbpo_cartpole \
        dynamics_model=gaussian_mlp_ensemble \
        ++device=cpu \
        ++overrides.sac_batch_size=256 \
        ++overrides.validation_ratio=0.2 \
        ++dynamics_model.activation_fn_cfg._target_=torch.nn.ReLU 
      ```
- Execute same command as in `.github/workflows/ci.yml` 
   - Execute dependencies install via requirement file
   - Lint with flake8
   - Lint with mypy
   - Check format with black


## Checklist




- [x] The documentation is up-to-date with the changes I made.
- [x] I have read the [**CONTRIBUTING**](../../CONTRIBUTING.md) document and completed the CLA (see **CONTRIBUTING**).
- [x] All tests passed, and additional code has been covered with new tests.
